File: 509-fibonacci-number/509-fibonacci-number.py
class Solution:
    def fib(self, n: int) -> int:
        def helper(n, memo):
            if n < 2: return n
            if n in memo: return memo[n]
            memo[n] = helper(n-1, memo) + helper(n-2, memo)
            return memo[n]
        
        return helper(n,{})

File: 509-fibonacci-number/509-fibonacci-number.cpp
class Solution {
public:
    map<int,int>cache;
    int fib(int n) {
        if (n < 2)
            return n;
        if (cache[n])
            return cache[n];
        cache[n] = fib(n-1) + fib(n-2);
        return cache[n];
    }
};